Revert "chapter 3 and 4 done"

abdul-moize opened the pull request:
Reverts abdul-moize/learn-python#1

arslanashraf7 commented on src/standard_libraries/test_re.py:
```diff
@@ -23,3 +23,8 @@ def test_re():
     # When only simple capabilities are needed, string methods are preferred because they are
     # easier to read and debug:
     assert 'tea for too'.replace('too', 'two') == 'tea for two'
+    regex = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
```
Could you do the below things here:

1. Can you format the regex something like [verbose formatting](https://diveintopython3.problemsolving.io/regular-expressions.html#verbosere)
2. Can we do something, If we are to introduce case-insensitivity?
3. Can we add one more test for a regex that would match Employee Card numbers? The format of the card number is (XX-XXXXX-44-XX)  -- Here X could be any number.

abdul-moize replied:
1. i have converted the regex from raw format to verbose format
2. we can remove the 'A-Z' to make it case sensitive. i have updated the regex to only accept lower case now
3. i have added this as well

arslanashraf7 commented on src/files/test_file_reading.py:
```diff
@@ -59,7 +60,7 @@ def test_files_open():
 
     # Open file using with.
-    with open('src/files/multi_line_file.txt', 'r') as file:
+    with open(path + 'multi_line_file.txt', 'r') as file:
```
Could you also try opening this fine in binary mode `rb`?

abdul-moize replied:
done

arslanashraf7 commented on src/files/test_file_reading.py:
```diff
@@ -40,9 +40,10 @@ def test_files_open():
 
     # Open files without using 'with' statement.
-    file = open('src/files/multi_line_file.txt', 'r')
+    file = open(path + 'multi_line_file.txt', 'r')
```
How about we do this file path generation with `.format`? Could you try it? You might need to update the path declaration.

abdul-moize replied:
done

arslanashraf7 commented on src/standard_libraries/test_json.py:
```diff
@@ -34,5 +34,8 @@ def test_json():
+    my_json_array = [{'name': 'amk', 'marks': 50}, {'name': 'hamza', 'marks': 30}]
+    assert my_json_array[0]['marks'] == 50
+    # notice the double quotes inside the json object
+    assert json.dumps(my_json_array) == '[{"name": "amk", "marks": 50}, {"name": "hamza", "marks": 30}]'
```
How about we create another test where we generate a json, something like:

`{
"name":  "Some Name",
"Age": 23,
"grades": [{
"title": "Science",
"total": 100,
"passing": 80,
"obtained": 85
},{
"title": "Math",
"total": 100,
"passing": 80,
"obtained":70
}]}
`

abdul-moize replied:
added a test case

arslanashraf7 reviewed (changes requested):
I've added some general suggestions, Along with them can you please add some function to test how 'pass' works?

abdul-moize commented:
> I've added some general suggestions, Along with them can you please add some function to test how 'pass' works?

i have updated the changes and also added a use case for pass. kindly check
